src/transports/ws.rs

```diff
@@ -0,0 +1,447 @@
+                                    }
+                                }
+                                Err(err) => {
+                                    debug!("accept websocket stream err: {:?}", err);
+                                }
+                            },
+                        }
+                    });
+                    self.poll_pending(cx)
+                }
+                Err(err) => {
+                    debug!("stream get peer address error: {:?}", err);
+                    Poll::Pending
+                }
+            },
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct WsStream {
+    inner: WebSocketStream<TcpStream>,
+    recv_buf: Vec<u8>,
+    pending_ping: Option<Vec<u8>>,
+    already_send_close: bool,
+}
+
+impl WsStream {
+    fn new(inner: WebSocketStream<TcpStream>) -> Self {
+        WsStream {
+            inner,
+            recv_buf: Vec::new(),
+            pending_ping: None,
+            already_send_close: false,
+        }
+    }
+
+    fn respond_ping(&mut self, cx: &mut Context) -> io::Result<()> {
+        if self.already_send_close {
+            return Ok(());
+        }
+        match self.pending_ping.take() {
+            Some(data) => {
+                let mut sink = Pin::new(&mut self.inner);
+                match sink.as_mut().poll_ready(cx) {
+                    Poll::Ready(Ok(_)) => {
+                        sink.as_mut()
+                            .start_send(Message::Pong(data))
+                            .map_err::<io::Error, _>(|e| {
+                                debug!("send error: {:?}", e);
+                                Into::into(io::ErrorKind::BrokenPipe)
+                            })?;
+                        let _ignore =
+                            sink.as_mut().poll_flush(cx).map_err::<io::Error, _>(|e| {
+                                debug!("flush error: {:?}", e);
+                                Into::into(io::ErrorKind::BrokenPipe)
+                            })?;
+                        Ok(())
+                    }
+                    Poll::Pending => {
+                        self.pending_ping = Some(data);
+                        Ok(())
+                    }
+                    Poll::Ready(Err(_)) => Err(Into::into(io::ErrorKind::BrokenPipe)),
+                }
+            }
+            None => Ok(()),
+        }
+    }
+
+    #[inline]
+    fn drain(&mut self, buf: &mut [u8]) -> usize {
+        // Return zero if there is no data remaining in the internal buffer.
+        if self.recv_buf.is_empty() {
+            return 0;
+        }
+
+        // calculate number of bytes that we can copy
+        let n = ::std::cmp::min(buf.len(), self.recv_buf.len());
+
+        // Copy data to the output buffer
+        buf[..n].copy_from_slice(self.recv_buf[..n].as_ref());
+
+        // drain n bytes of recv_buf
+        self.recv_buf = self.recv_buf.split_off(n);
```

Comment:
```suggestion
buf[..n].copy_from_slice(self.recv_buf.drain(..n).as_slice());
```

Reply:
I will use another PR to upgrade the toolchain and modify the related code